File: python/yacht/yacht.py
"""Score a game of Yacht."""
import collections


def sum_count(count: collections.Counter) -> int:
    return sum(val * num for val, num in count.items())


def face(val: int, count: collections.Counter) -> int:
    return val * count[val]


CATEGORIES = [
    ("ONES", None, lambda count: face(1, count)),
    ("TWOS", None, lambda count: face(2, count)),
    ("THREES", None, lambda count: face(3, count)),
    ("FOURS", None, lambda count: face(4, count)),
    ("FIVES", None, lambda count: face(5, count)),
    ("SIXES", None, lambda count: face(6, count)),
    ("LITTLE_STRAIGHT", (lambda count: set(count) == {1, 2, 3, 4, 5}), 30),
    ("BIG_STRAIGHT", (lambda count: set(count) == {2, 3, 4, 5, 6}), 30),
    ("FULL_HOUSE", (lambda count: set(count.values()) == {2, 3}), sum_count),
    (
        "FOUR_OF_A_KIND",
        (lambda count: count.most_common(1)[0][1] >= 4),
        (lambda count: 4 * count.most_common(1)[0][0]),
    ),
    ("YACHT", (lambda count: len(set(count)) == 1), 50),
    ("CHOICE", None, sum_count),
]
for name, condition, value in CATEGORIES:
    globals()[name] = (condition, value)


def score(dice: list[int], category) -> int:
    """Return the score of a "hand" in Yacht."""
    counter = collections.Counter(dice)
    condition, value = category
    if condition is None or condition(counter):
        return value(counter) if callable(value) else value
    return 0

File: go/yacht/yacht.go
// Package yacht scores a roll in Yacht.
package yacht

// I could map the string directly to a function but I opted
// to make use of an interface that has a .score method.
// This keeps the map and code easier to read, though longer.
var scorer = map[string]turn{
	"ones":            singles{1},
	"twos":            singles{2},
	"threes":          singles{3},
	"fours":           singles{4},
	"fives":           singles{5},
	"sixes":           singles{6},
	"full house":      fullHouse{},
	"four of a kind":  fourOfAKind{},
	"little straight": straight{6},
	"big straight":    straight{1},
	"choice":          choice{},
	"yacht":           yacht{},
}

// turn encapsulates information needed to score a turn.
type turn interface {
	score(dice []int) int
}

type singles struct {
	num int
}
type fullHouse struct{}
type fourOfAKind struct{}
type straight struct {
	missing int
}
type choice struct{}
type yacht struct{}

func count(dice []int) map[int]int {
	m := make(map[int]int)
	for _, d := range dice {
		m[d]++
	}
	return m
}

func sum(dice []int) (s int) {
	for _, d := range dice {
		s += d
	}
	return s
}

func (t singles) score(dice []int) (score int) {
	for _, d := range dice {
		if d == t.num {
			score += t.num
		}
	}
	return score
}

func (t fullHouse) score(dice []int) int {
	m := count(dice)
	if len(m) == 2 && (m[dice[0]] == 2 || m[dice[0]] == 3) {
		return sum(dice)
	}
	return 0
}

func (t fourOfAKind) score(dice []int) int {
	m := count(dice)
	if m[dice[0]] >= 4 {
		return 4 * dice[0]
	}
	if m[dice[1]] >= 4 {
		return 4 * dice[1]
	}
	return 0
}

func (t straight) score(dice []int) int {
	m := count(dice)
	if len(m) == 5 && m[t.missing] == 0 {
		return 30
	}
	return 0
}

func (t choice) score(dice []int) int {
	return sum(dice)
}

func (t yacht) score(dice []int) int {
	if len(count(dice)) != 1 {
		return 0
	}
	return 50
}

// Score scores a hand of dice and a category.
func Score(dice []int, category string) int {
	return scorer[category].score(dice)
}
